c/tskit/genotypes.h

```diff
@@ -34,46 +34,150 @@ extern "C" {
 
 #define TSK_ISOLATED_NOT_MISSING (1 << 1)
 
+/**
+@brief A variant at a specfic site.
+
+@rst
+Used to generate the genotypes for a given set of samples at a given
+site.
+@endrst
+*/
 typedef struct {
+    /** @brief Unowned reference to the tree sequence of the variant */
     const tsk_treeseq_t *tree_sequence;
+    /** @brief The site this variant is currently decoded at*/
     tsk_site_t site;
+    /** @brief Tree used by the variant to decode genotypes */
     tsk_tree_t tree;
+    /** @brief Array of allele strings that the genotypes of the variant refer to */
     const char **alleles;
+    /** @brief Lengths of the allele strings */
     tsk_size_t *allele_lengths;
+    /** @brief Length of the allele array */
     tsk_size_t num_alleles;
     tsk_size_t max_alleles;
+    /** @brief If True the genotypes of isolated nodes have been decoded to the "missing"
+     * genotype If False they set to the ancestral state (in the absence of mutations
+     * above them)*/
     bool has_missing_data;
+    /** @brief Array of genotypes for the current site */
     int32_t *genotypes;
-
+    /** @brief Number of samples */
     tsk_size_t num_samples;
-    tsk_size_t num_sites;
-    tsk_id_t *samples;                /* samples being used */
-    const tsk_id_t *sample_index_map; /* reverse index map being used */
+    /** @brief Array of sample ids used*/
+    tsk_id_t *samples;
+
+    const tsk_id_t *sample_index_map;
     bool user_alleles;
     char *user_alleles_mem;
     tsk_id_t *traversal_stack;
     tsk_flags_t options;
-    // private: the following data members are not intended to be used externally
-    tsk_id_t *alt_samples; /* alternative subset of samples to use */
+    tsk_id_t *alt_samples;
     tsk_id_t *alt_sample_index_map;
 
 } tsk_variant_t;
 
+/* All vargen related structs and methods were deprecated in C API v1.0 */
 typedef struct {
     const tsk_treeseq_t *tree_sequence;
     tsk_id_t site_index;
     tsk_variant_t variant;
 } tsk_vargen_t;
 
+/**
+@defgroup VARIANT_API_GROUP Variant API for obtaining genotypes.
+@{
+*/
+
+/**
+@brief Initialises the variant by allocating the internal memory and
+initialising the internal tree
+
+@rst
+This must be called before any operations are performed on the variant.
+See the :ref:`sec_c_api_overview_structure` for details on how objects
+are initialised and freed.
+@endrst
+
+@param self A pointer to an uninitialised tsk_variant_t object.
+@param tree_sequence A pointer to the tree sequence from which this variant
+will decode genotypes. No copy is taken, so this tree sequence must persist
+for the lifetime of the variant.
+@param samples An array of node ids of the samples that are to have their genotypes
```

Comment:
Need to document the NULL case here (we use the tree sequences's sample)

Reply:
https://github.com/tskit-dev/tskit/pull/2186/commits/9665c6fe0cfce938572f0b045a5d47dc1c9dce68